### PR TITLE
[SPARK-10688][ML][PySpark] Python API for AFTSurvivalRegression

### PR DESCRIPTION
**Pull request by vectorijk**
Implement Python API for AFTSurvivalRegression

### CODE REVIEW

**Comment by mengxr** [comment]
ok to test

**Comment by mengxr** [comment]
@GayathriMurali Could you post comments inline? You can select the "Files changed" tab and make inline comments.

File: python/pyspark/ml/regression.py [hunk 1]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
     .. versionadded:: 1.4.0
     """
 
+@inherit_doc
+class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,
+                            HasFitIntercept, HasMaxIter, HasTol):
+    """
+    `https://en.wikipedia.org/wiki/Accelerated_failure_time_model`
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: It would be nice to put the link in "see also" or embed it in the doc. The first sentence is usually a summary of the class.

(@GayathriMurali  This is an inline comment.)

File: python/pyspark/ml/regression.py [hunk 2]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+        self._java_obj = self._new_java_obj(
+            "org.apache.spark.ml.regression.AFTSurvivalRegression", self.uid)
+        self.censorCol = \
+            Param(self,  "censorCol",
+                  "censor column name")
+        self.quantileProbabilities = \
+            Param(self, "quantileProbabilities",
+                  "quantile probabilities array" +
+                  ", array is not empty and every probability is in range [0,1]")
+        self.quantilesCol = Param(self, "quantilesCol",
```
Comment (GayathriMurali) [inline python/pyspark/ml/regression.py]: Add comment to indicate what quantilesCol is

File: python/pyspark/ml/regression.py [hunk 3]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+                                  "quantiles column name")
+        self._setDefault(censorCol="censor",
```
Comment (GayathriMurali) [inline python/pyspark/ml/regression.py]: Add comment to indicate what quantile probabilities is and where it gets its value from -"This column will output quantiles of corresponding quantileProbabilities if it is set."

**Comment by GayathriMurali** [comment]
@mengxr Thanks for the input. I did add couple of inline comments, but instead of using file changed tab, i double clicked on the commit and added it. Made changes now

File: python/pyspark/ml/regression.py [hunk 4]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+        Param(Params._dummy(), "quantileProbabilities",
+              "quantile probabilities array" +
+              ", array is not empty and every probability is in range [0,1]")
+    quantilesCol = Param(Params._dummy(), "quantilesCol",
+                         "quantiles column name")
+
+    @keyword_only
+    def __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                 quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: We need to use `None` for quantileProbabilities. Otherwise, the default value is mutable and shared across all `AFTSurvivalRegression` instances. See https://issues.apache.org/jira/browse/SPARK-8678 and its associated PR for details.

File: python/pyspark/ml/regression.py [hunk 5]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+                 quantilesCol=None):
+        """
+        __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: Append `\` at the end to generate API doc correctly. You can run `make html` under `python/` to generate HTML docs locally and verify the output.

File: python/pyspark/ml/regression.py [hunk 6]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                 quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: Please keep the default array here instead of `None` because this only affects the doc.

File: python/pyspark/ml/regression.py [hunk 7]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+                 quantilesCol=None):
+        """
+        super(AFTSurvivalRegression, self).__init__()
+        self._java_obj = self._new_java_obj(
+            "org.apache.spark.ml.regression.AFTSurvivalRegression", self.uid)
+        self.censorCol = \
+            Param(self,  "censorCol",
+                  "censor column name")
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: 662-664 should fit into a single line

File: python/pyspark/ml/regression.py [hunk 8]
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+        self.setParams(**kwargs)
+
+    @keyword_only
+    @since("1.6.0")
+    def setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction",
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                 quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
+                 quantilesCol=None):
+        """
+        setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction",
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: Same about `\` at the end and the issue with mutable default value.

**Comment by vectorijk** [comment]
@mengxr @GayathriMurali Thanks so so much for code review! 
I cleaned up the code~

File: python/pyspark/ml/regression.py [hunk 9]
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+                         "corresponding quantileProbabilities if it is set.")
+
+    @keyword_only
+    def __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                 quantileProbabilities=None, quantilesCol=None):
+        """
+        __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction", \
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor", \
+                 quantileProbabilities=None, quantilesCol=None):
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: keep the default value of `quantileProbabilities` in docstring

File: python/pyspark/ml/regression.py [hunk 10]
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+
+    @keyword_only
+    @since("1.6.0")
+    def setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction",
+                  fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                  quantileProbabilities=None, quantilesCol=None):
+        """
+        setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction", \
+                  fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor", \
+                  quantileProbabilities=None, quantilesCol=None):
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: ditto

File: python/pyspark/ml/regression.py [hunk 11]
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
     """
 
 
+@inherit_doc
+class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,
+                            HasFitIntercept, HasMaxIter, HasTol):
+    """
+    Accelerated Failure Time(AFT) Model Survival Regression
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: space before `(` and add a period to the end.

File: python/pyspark/ml/regression.py [hunk 12]
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+
+    Fit a parametric AFT survival regression model based on the Weibull distribution
+    of the survival time.
+    see also, `https://en.wikipedia.org/wiki/Accelerated_failure_time_model`
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: use Sphinx `seealso` http://sphinx-doc.org/markup/para.html#directive-seealso Please generate the HTML doc and test the link.

File: python/pyspark/ml/regression.py [hunk 13]
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+
+    # a placeholder to make it appear in the generated doc
+    censorCol = Param(Params._dummy(), "censorCol",
+                      "censor column name. The value of this column could be 0 or 1. " +
+                      "If the value is 1, it means the event has occurred i.e. " +
+                      "uncensored; otherwise censored.")
+    quantileProbabilities = \
+        Param(Params._dummy(), "quantileProbabilities",
+              "quantile probabilities array. Values of the quantile probabilities array " +
+              "should be in the range [0, 1] and the array should be non-empty.")
```
Comment (mengxr) [inline python/pyspark/ml/regression.py]: The range should be `(0, 1)` instead of `[0, 1]`. Btw, it would be better to keep the doc in-sync with Scala. We can do that in a separate PR.

**Comment by mengxr** [comment]
LGTM except some minor inline comments

File: python/pyspark/ml/regression.py (hunk 13 above)
Comment (vectorijk) [inline python/pyspark/ml/regression.py]: I looked into [params.scala#L160](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/param/params.scala#L160). inRange() which uses inclusive be default. Should we revise the doc or code in Scala?

File: python/pyspark/ml/regression.py (hunk 12 above)
Comment (vectorijk) [inline python/pyspark/ml/regression.py]: okay! I changed it and tested it locally.

File: python/pyspark/ml/regression.py (hunk 13 above)
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: @mengxr @vectorijk  I think the range should be ```[0, 1]``` which is consistent with ```suvreg``` in R.

File: python/pyspark/ml/regression.py [hunk 14]
```diff
@@ -609,6 +611,150 @@ class GBTRegressionModel(TreeEnsembleModels):
...
+    def setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction",
+                  fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                  quantileProbabilities=None, quantilesCol=None):
+        """
+        setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction", \
+                  fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor", \
+                  quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99], \
+                  quantilesCol=None):
+        """
+        kwargs = self.__init__._input_kwargs
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: I think it's need special handling for ```quantileProbabilities``` here.
Because that if users did not set ```quantileProbabilities```, it will be set with ```None``` which is not default value ```[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]``` in the current route.
You can refer the [```layers```](https://github.com/apache/spark/blob/master/python/pyspark/ml/classification.py#L833) of MultilayerPerceptronClassifier.

Comment (vectorijk) [inline python/pyspark/ml/regression.py]: @mengxr @yanboliang changed that.
Plus, I added two methods in AFTSurvivalRegressionModel.

File: python/pyspark/ml/regression.py (hunk 13 above)
Comment (mengxr) [inline python/pyspark/ml/regression.py]: https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala#L282

~~~scala
q => lambda * math.exp(math.log(-math.log(1 - q)) / k)
~~~

If q = 0 or 1, the result is NA.

Comment (vectorijk) [inline python/pyspark/ml/regression.py]: changed that in doc

**Comment by vectorijk** [comment]
@mengxr  Could you review again cause I have changed it all?

**Comment by mengxr** [comment]
LGTM. Merged into master. I sent https://github.com/apache/spark/pull/9001 for a minor bug fix. Thanks!

**Comment by vectorijk** [comment]
@mengxr  Thanks so much for reviewing! It is my first PR to spark community. Excited!
P.S As you mentioned before, range of quantile probabilities is (0,1) instead of [0,1]. Should we create a PR for changing that in corresponding scala code?

**Comment by mengxr** [comment]
Yes. We should fix it. Could you make a JIRA and submit a PR? Thanks!

**Comment by vectorijk** [comment]
Sure
